### ISSUE TITLE
Dream7B的评估结果似乎是以AR的方式生成的？

### ISSUE THREAD

**Issue by duterscmy** [HKUNLP/Dream#72]
Hi 我正在复现论文中的结果，发现一个问题：

对于Dream7B-instruct 7B，脚本中默认的配置的temperature=0.1，当我将temperature改为0.0时，在GSM8K的准确率发生了显著下降（77%->58%)

这看起来并不正常，所以我在temperature=0.1的设置下，打印了推理过程中每个step时，所有mask token的置信度，然而，所有位置的置信度是相同的。这导致topk永远选择最左侧的token进行推理。

这个问题应该是由于generation_utils.py中sample_token函数的实现导致的，当设置温度>0时，返回的置信度是被温度影响的。

`def sample_tokens(logits, temperature=0.0, top_p=None, top_k=None, margin_confidence=False, neg_entropy=False):

    if temperature > 0:
        **logits = logits / temperature**
    if top_p is not None and top_p < 1:
        logits = top_p_logits(logits, top_p)
    if top_k is not None:
        logits = top_k_logits(logits, top_k)
    probs = torch.softmax(logits, dim=-1)

    if temperature > 0:
        try:
            **x0 = dists.Categorical(probs=probs).sample()
            confidence = torch.gather(probs, -1, x0.unsqueeze(-1)).squeeze(-1)**
        except:
            confidence, x0 = probs.max(dim=-1)
    else:
        print("temper=0.0")
        confidence, x0 = probs.max(dim=-1)
    
    if margin_confidence:
        sorted_probs, _ = torch.sort(probs, dim=-1, descending=True)
        # Extract top1 and top2 probabilities
        top1_probs = sorted_probs[:, 0] 
        top2_probs = sorted_probs[:, 1] 
        # Calculate confidence as top1 - top2
        confidence = top1_probs - top2_probs 
    
    if neg_entropy:
        epsilon = 1e-10
        **log_probs = torch.log(probs + epsilon)
        confidence = torch.sum(probs * log_probs, dim=-1)**
        print(confidence)
    return confidence, x0`

所以在论文中汇报的结果中， temperature=0.1的结果，实际是以AR的方式从左到右生成的，并非是diffusion generate得到？
